### DISCUSSION TITLE
CZI: Enable LaTeX

### DISCUSSION THREAD

**Discussion by IuryPiva** [bokeh/bokeh#11191] [General]
I'm starting this discussion to collect ideas and experiences on the optimal way of developing a class specifically for converting text into LaTex using MathJax. 

I've added an example of what would be needed to render it as a SVG in the AxisLabels, calculating its dimensions here #11128 

Some notes on converting tex to LaTex SVG using MathJax library:
 - It outputs an HTMLElementCollection with a SVGElement children.
 - Its size changes so positioning must be calculated based on the dimensions of the rendered SVG
 - MathJax library does not need to be included if there is no LaTex to be rendered, saving some space.
 - Loading an image is async, and TextBox is not.

**Comment by tcmetzger** [comment]
As a basis for this discussion, here are some ideas that @bryevdv, @IuryPiva and I have been talking about:

On the Python side, we will need those two classes:

1. class **`MathJaxConfig`** which has several individual properties mirroring MathJax' configuration properties (such as `font_cache = "local"`)
2. class **`MathString`** which has a `text` property (accepting a String) and a `config` property (accepting a `MathJaxConfig` object / the individual config properties)

Having a MathJaxConfig object also enables us to make it themeable - i.e. have this as part of a theme and use the same configuration for all MathString objects in a document.

The MathJax library is only included and loaded if there are any MathString objects in a document. Otherwise MathJax is not included.

Draft for `MathJaxConfig`:

```
class MathJaxConfig:
    font_cache = String()  # Maybe should be Enum()?
    [etc.]
```

Draft for `MathString`:

```
class MathString(inherit from something): 
    config = Readonly(Instance(MathJaxConfig))
    text = Nullable(String)
```

Example for setting a plot title:
```
p = figure()
p.title.text = MathString("\mu x^2")
p.title.text.config.font_cache = "local"
```

There is one notable **exception**:
A **`Div`** object will have a flag (e.g. `Div.math_config`) that defaults to `None `and additionally also accepts either `True` (i.e. all text in the `Div`'s string is put through Mathjax with a default configuration) or a `MathJaxConfig` object (i.e. all text in the `Div`'s string is put through Mathjax with the configuration defined by the `MathJaxConfig` object).

**Tick labels** and **legend labels** will be a little more complicated:
- For tick labels: We will see if it makes sense to have a dedicated MathJax formatter (see [User Guide for other examples of formatters](https://docs.bokeh.org/en/latest/docs/user_guide/styling.html#userguide-styling-axes-tick-label-formats)).
- [Automatically grouped legend labels](https://docs.bokeh.org/en/latest/docs/user_guide/annotations.html#automatic-grouping-python-side) (`legend_group`) will not support MathJax support
- Axis labels when set with the plotting interface (e.g. `p = figure(x_axis_label="\mu x^2")`) will need to be compatible with the current mechanism that collects all glyphs with the same axis label to pass a list to the Legend object created by the plotting interface's convenience function.

**Comment by tcmetzger** [comment]
Part of the CZI grant work is focused on adding LaTeX capabilities to Bokeh. The purpose of this discussion is to collect and discuss issues and pull requests related to LaTeX work. More information about other work that is part of this grant is available at #11351 and #11361.

**Comment by tcmetzger** [comment]
Related issues for this segment of work are:

- [x] Add LaTeX rendering to elements #11355
- [x] Examples demonstrating LaTeX capabilities #11356
- [x] Update docs for LaTeX capabilities #11357
- [x] Tests for LaTeX capabilities #11358
- [x] PNG export for LaTeX #11359
